scanpy/tests/test_ingest.py

```diff
@@ -0,0 +1,145 @@
+import pytest
+import numpy as np
+import scanpy as sc
+from scanpy.preprocessing._simple import N_PCS
+
+from sklearn.neighbors import KDTree
+from umap import UMAP
+
+
+X = np.array([
+    [ 1. ,  2.5,  3. ,  5. ,  8.7],
+    [ 4.2,  7. ,  9. , 11. ,  7. ],
+    [ 5.1,  2. ,  9. ,  4. ,  9. ],
+    [ 7. ,  9.4,  6.8,  9.1,  8. ],
+    [ 8.9,  8.6,  9.6,  1. ,  2. ],
+    [ 6.5,  8.9,  2.2,  4.5,  8.9]
+])
+
+T = np.array([
+    [2. , 3.5, 4. , 1. , 4.7],
+    [3.2, 2. , 5. , 5. , 8. ]
+])
+
+
+@pytest.fixture
+def adatas():
+    pbmc = sc.datasets.pbmc68k_reduced()
+    n_split = 500
+    adata_ref = sc.AnnData(pbmc.X[:n_split, :], obs=pbmc.obs.iloc[:n_split])
+    adata_new = sc.AnnData(pbmc.X[n_split:, :])
+
+    sc.pp.pca(adata_ref)
+    sc.pp.neighbors(adata_ref)
+    sc.tl.umap(adata_ref)
+
+    return adata_ref, adata_new
+
+
+def test_representation(adatas):
+    adata_ref = adatas[0].copy()
+    adata_new = adatas[1].copy()
+
+    ing = sc.tl.Ingest(adata_ref)
+    ing.transform(adata_new)
+
+    assert ing._use_rep == 'X_pca'
+    assert ing._obsm['rep'].shape == (adata_new.n_obs, N_PCS)
+    assert ing._pca_centered
+
+    sc.pp.pca(adata_ref, n_comps=30, zero_center=False)
+    sc.pp.neighbors(adata_ref)
+
+    ing = sc.tl.Ingest(adata_ref)
+    ing.transform(adata_new)
+
+    assert ing._use_rep  == 'X_pca'
+    assert ing._obsm['rep'].shape == (adata_new.n_obs, 30)
+    assert not ing._pca_centered
+
+    sc.pp.neighbors(adata_ref, use_rep='X')
+
+    ing = sc.tl.Ingest(adata_ref)
+    ing.transform(adata_new)
+
+    assert ing._use_rep == 'X'
+    assert ing._obsm['rep'] is adata_new.X
```

Comment:
In general, I think we should test the functionality instead of the implementation. For example, instead of checking `assert not ing._pca_centered`, could we check if the initial PCA wasn't centered, the new data wasn't centered before it was transformed?

Reply:
Do we need testing that the initial pca wasn't centered here?
I mean it is specified as `sc.pp.pca(adata_ref, n_comps=30, zero_center=False)` a few lines above, so it should be centered.

Reply:
When you check the value of `_pca_centered` you’re checking some internal state, not what computation happened. I think this should be tested more like this (this is basically pseudo code, since I’m not at a computer):

```python
refc = sc.pp.pca(adata, zero_center=True, copy=True)
ref = sc.pp.pca(adata, zero_center=False, copy=True)

# Not sure if I have these calls exactly right
# The intent is the `new` objects have been projected into the ref objects pca space
newc = sc.tl.ingest(adata, refc, embedding_method=“pca”, return_joint=False)
new = sc.tl.ingest(adata, ref, embedding_method=“pca”, return_joint=False)

assert np.allclose(newc.obsm[“X_pca”], refc.obsm[“X_pca”])
assert np.allclose(new.obsm[“X_pca”], ref.obsm[“X_pca”])
assert not np.allclose(new.obsm[“X_pca”], newc.obsm[“X_pca”])
```

You could alternatively use an oracle here, like scikit-learn’s PCA transformer.

Checking the value of `_pca_centered` doesn’t tell us whether these cases are true. Additionally, testing the behavior of the method rather than internal construction means we’re free to change the internals without having to rewrite tests.

This is basically the “test behavior, not implementation” mantra.